### PR TITLE
Migrate torch_export_python to ShapesSpec API

### PR DESCRIPTION
**Pull request by laithsakka**
Summary:
Switch torch_export_python's dynamic-shapes surface from the legacy
`torch.export.Dim` + dict-of-tuples form to the new
`torch.fx.experimental.dynamic_spec.ShapesSpec` API.

Public surface:
  Before:  dynamic_shapes={"x": (Dim("B"), Dim.STATIC)}

  After:   
dynamic_shapes={"x": TensorSpec([ShapeVar("B"), STATIC])} or
dynamic_shapes={"x": TensorSpec([ShapeVar("B"), 128])}

The new API is unbacked first, (no hidden silent specializations, it allows 
passing assumptions to specialize generated kernels). 

dynamic_shapes=ShapesSpec(
        params={
            "x": TensorSpec([B, STATIC, STATIC]),
        },
        assumptions = [B%16==0]
)

Snapshot specs (rms_norm, xsa_fwd), README, and end-to-end tests
(test_export.py) updated to the new API.

Test Plan:
added tests
run exisiting tests

Differential Revision: D109085691

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Review automatically exported from Phabricator review in Meta.
